feat(auth): add auth samples

FrodoTheTrue opened the pull request.

quartzmo reviewed (changes requested):
Please rename the following files to use snake-case:

* `auth/idTokenFromImpersonatedCredentials.go`
* `auth/idTokenFromMetadataServer.go`
* `auth/idTokenFromServiceAccount.go`
* `auth/verifyGoogleIdToken.go`

quartzmo commented on auth/idTokenFromImpersonatedCredentials.go:
```diff
@@ -0,0 +1,73 @@
+	// delegates: The chained list of delegates required to grant the final accessToken.
+	// For more information, see:
+	// https://cloud.google.com/iam/docs/create-short-lived-credentials-direct#sa-credentials-permissions
+	// Delegate is NOT USED here.
```
"Delegate is NOT USED here."

What does this mean?

FrodoTheTrue replied:
Moved to the request params

quartzmo commented on auth/idTokenFromMetadataServer.go:
```diff
@@ -0,0 +1,59 @@
+	// to the target audience.
+	_, err = ts.Token()
+	if err != nil {
+		return fmt.Errorf("enable to receive token: %v", err)
```
"enable"? Maybe meant: "failed" ?

quartzmo commented on auth/idTokenFromServiceAccount.go:
```diff
@@ -0,0 +1,66 @@
+	// to the target audience.
+	_, err = ts.Token()
+	if err != nil {
+		return fmt.Errorf("enable to receive token: %v", err)
```
"enable"? Maybe meant: "failed" ?

quartzmo commented on auth/idTokenFromImpersonatedCredentials.go:
```diff
@@ -0,0 +1,73 @@
+	// to the target audience.
+	_, err = ts.Token()
+	if err != nil {
+		return fmt.Errorf("enable to receive token: %v", err)
```
"enable"? Maybe meant: "failed" ?

quartzmo commented on auth/authenticate_implicit_with_adc.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+// authenticateImplicitWithAdc auto-detects credentials
+// and prints storage buckets by authenticating with ADC.
```
Maybe state the full-length term Application Default Credentials, or else provide a link?

piaxc replied:
+1, maybe just
authenticateImplicitWithAdc uses Application Default Credentials
to automatically find credentials and authenticate.

quartzmo commented on auth/authenticate_explicit_with_adc.go:
```diff
@@ -0,0 +1,71 @@
+	"google.golang.org/api/option"
+)
+
+// authenticateExplicitWithAdc prints storage buckets by authenticating with ADC.
```
Maybe state the full-length term Application Default Credentials, or else provide a link?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+	buf := &bytes.Buffer{}
+
+	if err := authenticateExplicitWithAdc(buf); err != nil {
+		t.Errorf("authenticateExplicitWithAdc got err: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+	buf.Reset()
+
+	if err := authenticateImplicitWithAdc(buf, tc.ProjectID); err != nil {
+		t.Errorf("authenticateImplicitWithAdc got err: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+	buf.Reset()
+
+	if err := getIdTokenFromMetadataServer(buf, audience); err != nil {
+		t.Errorf("getIdTokenFromMetadataServer got err: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+	buf.Reset()
+
+	if err := getIdTokenFromServiceAccount(buf, os.Getenv("GOOGLE_APPLICATION_CREDENTIALS"), audience); err != nil {
+		t.Errorf("getIdTokenFromServiceAccount got err: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+
+	credentials, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		t.Errorf("failed to generate default credentials: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+
+	ts, err := idtoken.NewTokenSource(ctx, audience, option.WithCredentials(credentials))
+	if err != nil {
+		t.Errorf("failed to create NewTokenSource: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+
+	token, err := ts.Token()
+	if err != nil {
+		t.Errorf("failed to get ID token: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

quartzmo commented on auth/auth_test.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	if err := verifyGoogleIdToken(buf, token.AccessToken, audience); err != nil {
+		t.Errorf("verifyGoogleIdToken got err: %v", err)
```
Use `t.Fatalf` here because subsequent operation(s) depend on this one succeeding?

piaxc commented on auth/authenticate_explicit_with_adc.go:
```diff
@@ -0,0 +1,71 @@
+	// For more information on scopes to use,
+	// see: https://developers.google.com/identity/protocols/oauth2/scopes
+
+	// onstruct the Storage client.
```
typo, should be "Construct"

piaxc commented on auth/authenticate_implicit_with_adc.go:
```diff
@@ -0,0 +1,61 @@
+
+	ctx := context.Background()
+
+	// This snippet demonstrates how to list buckets.
```
Please remove, while strictly true, it's not really the point of the snippet.

piaxc reviewed (changes requested):
Thanks Fedor!

FrodoTheTrue commented:
@quartzmo could you please take a loot once more?
Also, I am receiving stange lint error ...

```
To fix this check, run "goimports -w . && find . -name go.mod -execdir go mod tidy \;"
HEAD detached at pull/2706/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   auth/auth_test.go
	modified:   auth/id_token_from_impersonated_credentials.go

no changes added to commit (use "git add" and/or "git commit -a")
Error: Process completed with exit code 1.
```

Not sure what to do with that

FrodoTheTrue commented:
@codyoss could you please help me with an error above?

codyoss commented:
Looks like you need to run goimports to format the files.

FrodoTheTrue commented:
@quartzmo could you please take a look once more?

codyoss commented on auth/id_token_from_impersonated_credentials.go:
```diff
@@ -0,0 +1,71 @@
+// getIdTokenFromImpersonatedCredentials uses a service account (SA1) to impersonate as
+// another service account (SA2) and obtain id token for the impersonated account.
+// To obtain token for SA2, SA1 should have the "roles/iam.serviceAccountTokenCreator" permission on SA2.
+func getIdTokenFromImpersonatedCredentials(w io.Writer, scope, targetAudience, impersonatedServiceAccount string) error {
```
Can we add a test for this?

FrodoTheTrue replied:
It's not clear how to test impersonated token creation, so the team decided for now to skip this in all languages.
Ex: https://github.com/googleapis/google-auth-library-python/blob/main/samples/cloud-client/snippets/snippets_test.py

codyoss commented on auth/id_token_from_impersonated_credentials.go:
```diff
@@ -0,0 +1,71 @@
+
+	// Construct the GoogleCredentials object which obtains the default configuration from your
+	// working environment.
+	credentials, err := google.FindDefaultCredentials(ctx, scope)
```
Not creating this type and passing this has the same affect, can you remove this from the example?

FrodoTheTrue replied:
It will work, but the goal of the samples is to show how to explicitly set up the credentials
Examples from other languages: https://github.com/googleapis/google-auth-library-python/blob/main/samples/cloud-client/snippets/idtoken_from_impersonated_credentials.py#L47

codyoss replied:
For these two can we add to the comment to say that this would be done implicitly or something to that nature? I think it is fine to be explicit. But if we don't say we are being verbose users are going to follow this pattern without knowing that there is a simpler way.

FrodoTheTrue replied:
I guess we have "authenticate explicit" example and it should be enough from my view.

Again, same for other languages (there are additional info): https://github.com/googleapis/google-auth-library-python/pull/1102/files#diff-96271f4091235622616d8e8fa7d45ddf6da62cd552544721dc10214a3562d57dR47

codyoss replied:
That is true there is a sample like that but think about it from a user point of view. I don't think that sample and this one will be shown on the same page, right? If there are not they don't have the context of the other sample. I get that this is how things were done in other langs too, but I wonder if it is right to do so there as well.

Whenever possible I think we should have users rely on our auto-detection logic and only in exceptional cases be explicit, when required. I will approve this, but might be worth talking to the team about.

codyoss commented on auth/id_token_from_metadata_server.go:
```diff
@@ -0,0 +1,59 @@
+
+	// Construct the GoogleCredentials object which obtains the default configuration from your
+	// working environment.
+	credentials, err := google.FindDefaultCredentials(ctx)
```
Same as previous file, I think this should just be removed as it is not needed to make the code work.
